llvm/lib/Target/AIE/AIEInterBlockScheduling.h

```diff
@@ -166,23 +166,77 @@ class Region {
   // The instruction that starts the next region, if any
   MachineInstr *ExitInstr = nullptr;
 
+  MachineBasicBlock *BB = nullptr;
+
+  /// Instructions that are already scheduled at the top.
+  /// Those should not be re-ordered by the scheduler.
+  ArrayRef<MachineBundle> TopFixedBundles;
+
+  /// Instructions that are already scheduled at the bottom.
+  /// Those should not be re-ordered by the scheduler.
+  ArrayRef<MachineBundle> BotFixedBundles;
+
 public:
   Region(MachineBasicBlock *BB, MachineBasicBlock::iterator Begin,
-         MachineBasicBlock::iterator End) {
-    for (auto It = Begin; It != End; ++It) {
+         MachineBasicBlock::iterator End,
+         ArrayRef<MachineBundle> TopFixedBundles,
+         ArrayRef<MachineBundle> BotFixedBundles)
+      : BB(BB), TopFixedBundles(TopFixedBundles),
+        BotFixedBundles(BotFixedBundles) {
+    MachineBasicBlock::iterator FreeBegin =
+        std::next(Begin, TopFixedBundles.size());
+    MachineBasicBlock::iterator FreeEnd =
+        std::prev(End, BotFixedBundles.size());
+
+    // Verify that all fixed instructions are at the right place in the MBB
+    assert(TopFixedBundles.empty() || Begin == BB->begin());
+    assert(TopFixedBundles.empty() ||
+           all_of(TopFixedBundles.back().Instrs, [FreeBegin](
+                                                     const MachineInstr *MI) {
+             return getBundleStart(MI->getIterator()) == std::prev(FreeBegin);
+           }));
+    assert(BotFixedBundles.empty() || End == BB->end());
+    assert(BotFixedBundles.empty() ||
+           all_of(BotFixedBundles.front().Instrs,
+                  [FreeEnd](const MachineInstr *MI) {
+                    return getBundleStart(MI->getIterator()) == FreeEnd;
+                  }));
+
+    for (auto It = FreeBegin; It != FreeEnd; ++It) {
       SemanticOrder.push_back(&*It);
```

Comment:
Maybe mention whether and where fixed instructions are added to the semantic order.